src/StrList.h

```diff
@@ -13,16 +13,76 @@
 
 #include "sbuf/forward.h"
 
+#include <iterator>
+
 class String;
+class StrList;
 
+/// Appends the given item to a delimiter-separated list in str.
 void strListAdd(String * str, const char *item, char del);
+
+/// Appends the given item to a delimiter-separated list in str.
+/// Use strListAdd(c-string) for c-string items with unknown length.
+void strListAdd(String &str, const SBuf &item, char delimiter = ',');
+
 int strListIsMember(const String * str, const SBuf &item, char del);
 int strListIsSubstr(const String * list, const char *s, char del);
+/// Iterates through delimiter-separated and optionally "quoted" list members.
+/// Follows HTTP #rule, including skipping OWS and empty members.
+/// \deprecated use a C++ range-based "for" loop with StrList instead
 int strListGetItem(const String * str, char del, const char **item, int *ilen, const char **pos);
```

Comment:
@yadij, after you [banned](https://github.com/squid-cache/squid/pull/481#discussion_r359084690) string views, StrList iterator naturally became very slow -- allocating, copying, and deleting an SBuf on every iteration. This comment (added before the ban) still deprecates the old ugly iteration interface in favor of this new range-based `for` loops, but the comment is now encouraging folks to introduce performance regressions!

I see two ways to resolve this conflict and finally merge this long-requested feature:

1. Introduce non-copyable string views that StrList can use to iterate efficiently. The callers will not be able to accidentally copy/store these views, reducing view risks to that of a C++ reference (which we are still allowed to use). I recommend this option because it moves us forward.

2. Remove StrList with all the advantages it offers for safer list header field iteration. Rewrite PR code to use the old strListGetItem(). The code will become uglier (we have started with that approach so I have seen it), but it will not introduce and encourage performance regressions.

Christos or I can do the legwork required for either change, but the call is yours. How do you want us to proceed?

Reply:
> @yadij, after you [banned](https://github.com/squid-cache/squid/pull/481#discussion_r359084690) string views, StrList iterator naturally became very slow -- allocating, copying, and deleting an SBuf on every iteration. This comment (added before the ban) still deprecates the old ugly iteration interface in favor of this new range-based `for` loops, but the comment is now encouraging folks to introduce performance regressions!

Do you have any data supporting this assertion? I can see why SBuf could be a little slower than any alternative, but not very slow
TBH I find the whole concept of StrList to be quite ugly, and that its use would be better served (as an API, not talking about performance here) by a std::list<SBuf>, either as a first-class citizen or as a cache to access the underlying multiply-referenced MemBlob. This is however out of scope for this PR

> I see two ways to resolve this conflict and finally merge this long-requested feature:
[...]
> Christos or I can do the legwork required for either change, but the call is yours. How do you want us to proceed?

I'll let this answer to @yadij , as he has more context than me

Reply:
> Alex: [After string views were banned], StrList iterator naturally became very slow -- allocating, copying, and deleting an SBuf on every iteration.

> Francesco: Do you have any data supporting this assertion? 

Allocating, copying, and deleting an SBuf on every iteration is how an SBuf-based StrList  works. StrListIterator::proceed() code clearly [shows](https://github.com/squid-cache/squid/blob/ae88e5b70a2b2009838a014aba331d67f33ce476/src/StrList.cc#L176) a temporary SBuf created for every iteration.

The (deleted) view-based iteration contained no buffer allocation, copying, or deleting. Zero. There is not much there to talk about as far as expense is concerned -- copying a pointer and an integer to a location on stack is pretty much it (and even that might be optimized out in some cases).

Do you need more "data" than that?

> I can see why SBuf could be a little slower than any alternative, but not very slow

You would have to define "very slow" for a meaningful discussion, but the performance difference I am talking about here is significant: Allocating, copying, and deleting an SBuf on every iteration is pretty much as expensive as it gets when iterating constant Strings. Frankly, it is difficult to imagine what can be _slower_ in this context. Allocating a buffer for every character?! (Rhetorical) Using views requires  _none_ of those expensive operations!

> TBH I find the whole concept of StrList to be quite ugly

If you are talking about the legacy StrList*() functions, then they are indeed very ugly. However, they have pretty much nothing to do with this discussion. This PR does not change them. This PR did try to deprecate the ugliest of the bunch -- strListGetItem(). The outcome depends on this change request.

If you are talking about the new StrList class (which is meant as strListGetItem replacement), then I think you are missing something because I cannot imagine a much simpler or nicer interface. Here it is in action (see the second line; the first line just gets the stored header field value):

```C++
const auto protocols = reply.header.getList(Http::HdrType::UPGRADE);
for (const auto &protocol: StrList(protocols)) {
    debugs(11, 5, "server accepted " << protocol);
    ...
}
```

I doubt there can be a much better interface for iterating lists stored in a String (which is what we have to iterate in this PR and many places in the official code).

> its use would be better served (as an API, not talking about performance here) by a std::list

If we start storing HTTP list elements such as Upgrade protocols and Cache-Control directives in std::list, then there will be no need to iterate them using StrList, of course -- std::list provides its own begin() and end() methods which allow for a similarly nice range-based iteration. However, discussing how HttpHeader and friends should store list elements in the future is outside this PR scope (not to mention changing that storage). This PR has to deal with String-based storage.

Hope this clarifies.

Reply:
> [Amos](https://github.com/squid-cache/squid/pull/481#issuecomment-632939242): As I said earlier avoiding the SBuf allocations in old code is one of the major reasons why MemBuf still exists.

That assertion is irrelevant to this change request [question](https://github.com/squid-cache/squid/pull/481#discussion_r428888186), so I am not going to debate its correctness.

> The StringView class definition was essentially this:

```C++
class StringView : private MemBuf
{
    StringView(char *s, int len) : stolen(1), buf(s), size(len), max_capacity(len), capacity(len) {}

    bool empty() const { return !hasContent(); }
    size_t size() const { return contentSize(); }
    const char *data() const { return content(); }
};
```

You did not ban StringView because of StringView implementation. This particular (incorrect) way to implement StringView is a red herring. Yes, one can abuse, say, MemBuf to implement StringView. That fact has nothing to do with the question posed by this change request. The question is _not_ about StringView _implementation_.

Reply:
To proceed with this PR I am referring to a compromise solution:
- Add a new method HttpHeader::StrList(Http::HdrType id, SBuf &out)
- Fix StrList to get SBuf input instead of String:
   ```StrList::StrList(const SBuf &s, const char delim = ',')```
- Provide an bool strListGetItem(const SBuf &str, char del, SBuf &item,  size_t pos) for use with StrList. The new method duplicates functionality and probably code, but I hope In the future the current strListGetItem will be removed.

The above needs some work, but I hope not a lot. I hope that the returned SBuf list items will not reallocate but just will point to the same blob.

In any case the way the HttpHeader works, already forces many strings copying and reallocations (should be fixed in the future to use SBuf instead of String). I hope if this part of code adds some re-allocations, is not a huge overhead, the requests with upgrade headers  I hope will not be many or not have many items in the Upgrade header.

Reply:
I see no good reason to duplicate HttpHeader::getList() (and strListAdd): If Amos bans string views even in the proposed form of temporary/non-copyable objects, then we should just use the existing strListGetItem() API and be done with it.

Yes, HttpHeader storage should be revised, but that revision should not be done using piecemeal at hoc approach. It deserves a serious dedicated effort and a stand-alone PR.

> I hope if this part of code adds some re-allocations, is not a huge overhead

There is no need to add re-allocations here so agreeing on whether they add huge, tolerable, reasonable, average, small, minor, negligible, or whatever overheads is not only impractical but unnecessary.

> the requests with upgrade headers I hope will not be many or not have many items in the Upgrade header.

The header iteration APIs in question are not Upgrade-specific so this point is rather moot. It is also debatable whether it is a good idea to sacrifice performance of Upgrade-rich environments because other environments cannot care less about Upgrade.

Sometimes, we have to make performance tradeoffs. There is no good reason to trade performance here AFAICT -- we can have a performant API for all users. It is Amos' call whether we use the old one or add a new one. The tradeoff here is code quality, not performance.

Reply:
My point with the "class StringView : private MemBuf" comment was not to add such a definition (please don't, extend MemBuf if anything). It was to show that MemBuf provides the API you need for the good-performance implementation. It will not be nice code to look at, but this StrList stuff is all on the chopping block for purge with Parser rewrites so looking good is irrelevant.

So if we really need the performance MemBuf is there, otherwise SBuf has the long-term API with temporary performance issues.

If you want me to pick up the work/cost of converting to MemBuf I can take that on.

Reply:
@yadij, I strongly disagree that MemBuf provides a suitable API for a proper string view implementation. I would block any PR that attempts to misuse MemBuf in such a way. Since I strive to apply the same principles to all PRs, I cannot accept your kind offer of implementing string views using MemBuf.

@chtsanti, please remove StrList and go back to using the old strListGetItem() API for iterating String lists. We went above and beyond to provide a much better solution, but I do not think this PR should be held hostage to our current inability to convince Amos that string views have legitimate use cases.

Reply:
The StrList/StrListIterator removed with the commit 9033506. 
I also committed a small patch for ProtocolView class in commit 65d3826

Please re-review.

Reply:
@chtsanti, your commit 65d3826 changes were rather clever, thank you!

I consider this change request addressed as of 1183d92.